### ISSUE TITLE
quick and multiple Ctrl+C interrupt for kubectl logs -f cause konnectivity-agent stuck

### ISSUE THREAD

**Issue by ydp** [kubernetes-sigs/apiserver-network-proxy#167]
Hi team,

I met a strange issue in our product that when I use `kubectl logs -f konnectivity-agent-xxx` to streaming logs, because there are many logs there, I did not wait it to finish and use `Ctrl+C` to interrupt the streaming, since it did not exit immediately like cat big file, I issued `Ctrl+C` at least 3 times as quick as I can. Then the streaming process exit, but konnectivity got stuck. This looks like a random behavior, but after several experiments, it actually can be reproduced in any of my setups.

konnectivity server/client version: v0.0.12
setup: one proxy-server on k8s master vm and one proxy-agent run as pod(one replica deployment).

The logs when `Ctrl+C` issued:
server:
```
I0121 06:53:12.616717       7 server.go:224] proxy request from client, userAgent [grpc-go/1.27.0]
I0121 06:53:12.616941       7 server.go:257] start serving frontend stream
I0121 06:53:12.617098       7 server.go:268] >>> Received DIAL_REQ
I0121 06:53:12.617118       7 backend_manager.go:170] pick agentID=a89e1d16-8908-4f83-bbc1-bea45ba34ea4 as backend
I0121 06:53:12.617168       7 server.go:290] >>> DIAL_REQ sent to backend
I0121 06:53:12.740584       7 server.go:308] >>> Received 24 bytes of DATA(id=195)
I0121 06:53:12.740750       7 server.go:324] >>> DATA sent to Backend
I0121 06:53:12.740818       7 server.go:294] >>> Received CLOSE_REQ(id=195)
I0121 06:53:12.740842       7 server.go:303] >>> CLOSE_REQ sent to backend
I0121 06:53:13.353224       7 server.go:308] >>> Received 24 bytes of DATA(id=194)
I0121 06:53:13.353335       7 server.go:324] >>> DATA sent to Backend
I0121 06:53:13.353347       7 server.go:294] >>> Received CLOSE_REQ(id=194)
I0121 06:53:13.353357       7 server.go:303] >>> CLOSE_REQ sent to backend
I0121 06:53:13.405226       7 server.go:308] >>> Received 24 bytes of DATA(id=189)
I0121 06:53:13.405460       7 server.go:324] >>> DATA sent to Backend
I0121 06:53:13.405484       7 server.go:294] >>> Received CLOSE_REQ(id=189)
I0121 06:53:13.405546       7 server.go:303] >>> CLOSE_REQ sent to backend
I0121 06:53:14.735760       7 server.go:308] >>> Received 24 bytes of DATA(id=192)
I0121 06:53:14.735980       7 server.go:324] >>> DATA sent to Backend
I0121 06:53:14.736004       7 server.go:294] >>> Received CLOSE_REQ(id=192)
I0121 06:53:14.736054       7 server.go:303] >>> CLOSE_REQ sent to backend
I0121 06:53:17.537537       7 server.go:224] proxy request from client, userAgent [grpc-go/1.27.0]
I0121 06:53:17.537740       7 server.go:257] start serving frontend stream
I0121 06:53:17.537861       7 server.go:268] >>> Received DIAL_REQ
I0121 06:53:17.538072       7 backend_manager.go:170] pick agentID=a89e1d16-8908-4f83-bbc1-bea45ba34ea4 as backend
I0121 06:53:17.538320       7 server.go:290] >>> DIAL_REQ sent to backend
```
client:
```
I0121 06:53:12.619855       1 client.go:280] received DIAL_REQ
I0121 06:53:12.680382       1 client.go:271] [tracing] recv packet, type: DIAL_REQ
I0121 06:53:12.680413       1 client.go:280] received DIAL_REQ
I0121 06:53:12.743029       1 client.go:271] [tracing] recv packet, type: DATA
I0121 06:53:12.743058       1 client.go:339] received DATA(id=195)
I0121 06:53:12.743077       1 client.go:271] [tracing] recv packet, type: CLOSE_REQ
I0121 06:53:12.743083       1 client.go:350] received CLOSE_REQ(id=195)
I0121 06:53:12.743088       1 client.go:306] close connection(id=195)
E0121 06:53:12.743274       1 client.go:419] conn write error: write tcp 192.168.113.19:52218->192.168.113.18:10349: use of closed network connection
I0121 06:53:12.743986       1 client.go:384] received 0 bytes from remote for connID[195]
W0121 06:53:12.744018       1 client.go:391] connection read error: read tcp 192.168.113.19:52218->192.168.113.18:10349: use of closed network connection
I0121 06:53:13.355882       1 client.go:271] [tracing] recv packet, type: DATA
I0121 06:53:13.355966       1 client.go:339] received DATA(id=194)
I0121 06:53:13.355997       1 client.go:271] [tracing] recv packet, type: CLOSE_REQ
I0121 06:53:13.356005       1 client.go:350] received CLOSE_REQ(id=194)
I0121 06:53:13.356011       1 client.go:306] close connection(id=194)
I0121 06:53:13.356206       1 client.go:384] received 0 bytes from remote for connID[194]
W0121 06:53:13.356218       1 client.go:391] connection read error: read tcp 192.168.113.19:52216->192.168.113.18:10349: use of closed network connection
E0121 06:53:13.356451       1 client.go:419] conn write error: write tcp 192.168.113.19:52216->192.168.113.18:10349: use of closed network connection
I0121 06:53:13.408583       1 client.go:271] [tracing] recv packet, type: DATA
I0121 06:53:13.408690       1 client.go:339] received DATA(id=189)
I0121 06:53:13.408809       1 client.go:271] [tracing] recv packet, type: CLOSE_REQ
I0121 06:53:13.408885       1 client.go:350] received CLOSE_REQ(id=189)
I0121 06:53:13.409302       1 client.go:306] close connection(id=189)
I0121 06:53:13.409049       1 client.go:413] [connID: 189] write last 24 data to remote
I0121 06:53:13.409572       1 client.go:384] received 0 bytes from remote for connID[189]
W0121 06:53:13.409620       1 client.go:391] connection read error: read tcp 192.168.113.19:52194->192.168.113.18:10349: use of closed network connection
I0121 06:53:14.738572       1 client.go:271] [tracing] recv packet, type: DATA
I0121 06:53:14.738614       1 client.go:339] received DATA(id=192)
I0121 06:53:14.738659       1 client.go:271] [tracing] recv packet, type: CLOSE_REQ
I0121 06:53:14.738665       1 client.go:350] received CLOSE_REQ(id=192)
I0121 06:53:14.738864       1 client.go:306] close connection(id=192)
I0121 06:53:14.739285       1 client.go:384] received 0 bytes from remote for connID[192]
W0121 06:53:14.739373       1 client.go:391] connection read error: read tcp 192.168.113.19:52212->192.168.113.18:10349: use of closed network connection
E0121 06:53:14.739612       1 client.go:419] conn write error: write tcp 192.168.113.19:52212->192.168.113.18:10349: use of closed network connection
I0121 06:53:17.541034       1 client.go:271] [tracing] recv packet, type: DIAL_REQ
I0121 06:53:17.541071       1 client.go:280] received DIAL_REQ
I0121 06:53:17.577462       1 client.go:271] [tracing] recv packet, type: DIAL_REQ
I0121 06:53:17.577503       1 client.go:280] received DIAL_REQ
I0121 06:53:17.890636       1 client.go:271] [tracing] recv packet, type: DIAL_REQ
...
...
```
After client side got stuck, it just print above last 2 lines of logs continuously. I could see there are 4 CLOSE_REQ (195,194,189,192) send through server to client, which looks like just my actions. I am not sure if this is helpful, but anyway the service 192.168.113.18:10349 is Antrea controller, we are using Antrea as CNI.

**Comment by ydp** [comment]
I found that the kubectl version in our version is v1.12.7, change to the latest version, the problem seems mitigate a lot.

**Comment by cheftako** [comment]
Any chance you can crank up the logging level (-v 9) and reproduce?

**Comment by ydp** [comment]
@cheftako  thanks for the reply, I added "--v=9" to both server and client side, it did not print out more logs than before, is this the right flag to add? 
and I updated both server and agent side to tag v0.0.14, the issue still reproduces.

**Comment by ydp** [comment]
After some investigation, I had several confusion, could anyone help answer?

1. Regarding above issue, when I interrupt one logs stream, is it normal that all other connections received CLOSE_REQ? 
2. I just noticed that in our k8s release(`1.19.6`), it's using konnectivity-client v.0.0.9, but we are actually using proxy-server and client v.0.0.12+, are this two versions compatible? Is it possible my issue caused by the version compatibility? why don't we bump the k-client version in k8s 1.19 branch?

Thanks

**Comment by cheftako** [comment]
> After some investigation, I had several confusion, could anyone help answer?
> 
> 1. Regarding above issue, when I interrupt one logs stream, is it normal that all other connections received CLOSE_REQ?

So there are 2 tunnel sections, 1 from KAS to Server and another from Server to Agent. We do not reuse the tunnel section from KAS to Server. However we do multiplex traffic on the tunnel sections from Server to Agent. It sounds like that segment may be improperly broken. Do you know if the Agent remains healthy through the whole exchange? If the Agent is crashing that could explain what you are seeing. Otherwise it could be an issue with the backend manager in the Server.

> 2. I just noticed that in our k8s release(`1.19.6`), it's using konnectivity-client v.0.0.9, but we are actually using proxy-server and client v.0.0.12+, are this two versions compatible? Is it possible my issue caused by the version compatibility? why don't we bump the k-client version in k8s 1.19 branch? This PR [kubernetes/kubernetes#94983](https://github.com/kubernetes/kubernetes/pull/94983) only merged into master branch.

As far as I know those version are compatible, but @Jefftree has been managing that and can confirm. Bumping the k-client version in any K8s branch requires a cherry pick. The release process discourages prophylactic cherry-picks and instead prefers we only cherry pick for known concrete bugs.

> 
> Thanks

**Comment by ydp** [comment]
After upgrade to v0.0.15 for k8s 1.20.5, the issue still exists.

**Comment by Jefftree** [comment]
Confirming that I can reproduce this problem. I'm not exactly sure about the mechanisms of how we get into the specified state, but the end result is that the agent is unable to proxy any more requests. 

Agent log show a stream of DIAL_REQs

```
I0507 19:34:50.096369       1 client.go:330] received DIAL_REQ
I0507 19:34:57.461235       1 client.go:330] received DIAL_REQ
I0507 19:35:06.905829       1 client.go:330] received DIAL_REQ
I0507 19:35:06.906342       1 client.go:330] received DIAL_REQ
I0507 19:46:46.480205       1 client.go:330] received DIAL_REQ
...
```

while still being able to communicate with the proxy server. The agent <-> server connection is still healthy and no health checks are failing. A successful DIAL_RSP is not sent back from the agent to the server so the client will hang 30s and timeout (apiserver limit). I'm suspecting there's an issue with the dial to the remote destination, probably related to https://github.com/kubernetes-sigs/apiserver-network-proxy/pull/179. 

Still need to do some more debugging to find the root cause.

**Comment by jkh52** [comment]
I can repro this. Out of curiosity I tested with binaries built to include https://github.com/kubernetes-sigs/apiserver-network-proxy/pull/179 to see if it helps, but it still occurs.

**Comment by Avatat** [comment]
I confirm, that problem exists on agent and server in version 0.0.17.

**Comment by caesarxuchao** [comment]
Summarized the root cause and solutions in this google drive [doc](https://docs.google.com/document/d/1lI1bPJg6hV_hSDW9k530GqnckVRNGn1-apQFx64tqjw/edit?resourcekey=0-P2PTM6V2ZoR_ERT9T3Fr6A#).

You need to join https://groups.google.com/g/kubernetes-sig-api-machinery or https://groups.google.com/g/kubernetes-sig-cloud-provider to view.

**Comment by ydp** [comment]
@caesarxuchao @cheftako @jkh52 Awesome work! This fix and following up changes definitely make this component more robust. Though you already tested this, I will double confirm it once we bump to the latest version.

I'm learning the design doc, since no comment permission, will just ask here, appreciate if anyone could help answer.

* Why the golang http lib sometime can or cannot consume the last piece of data, which one is the expected behavior? I mean is this a golang http lib defect?
* Does the last piece of data mean one packet or many data? Could we set flow control windows in GRPC somewhere to make the data always consumed by client?
* When the doc mentions track trace, it's retrieved with the latest pprof (#201) support, is it possible to make it enabled by default? It seems that kubelet has it enabled by default.
* Just a confirm, the `when ctrl-c is hit` section in the doc, the six step should be `then sends the CLOSE_RSP back to the proxy server`, right?
* As the solution 2 is mentioned, how much cost that the timer for each packet is? Because we might encounter some performance issue in using `kubectl` to execute test cases, which is not confirmed yet, but the symptom is that in our scale test, the test cases take more time than other product that does not the ANP. Since it only affects test cases themselves, not the feature it tests, I did not pay much attention on this yet.
